### PR TITLE
Expose TextureViewArray, SamplerArray and associated extensions to the C API

### PR DESCRIPTION
**Pull request by Linx145**
I've been using the wgpu-native bindings from C# and realised that one of the features I required was missing, namely texture view arrays/sampler arrays as specified here (https://wgpu.rs/doc/wgpu/enum.BindingResource.html), so I implemented the features and created this pull request. It also exposes TEXTURE_BINDING_ARRAY and SAMPLED_TEXTURE_AND_STORAGE_BUFFER_ARRAY_NON_UNIFORM_INDEXING as features to the C API via wgpu.h

I'm not a rust developer and have never written anything in rust before. This is also my first pull request so any feedback is appreciated!

### CODE REVIEW

File: examples/triangle/main.c [hunk 1]
```diff
@@ -5,6 +5,8 @@
 #include <stdio.h>
 #include <stdlib.h>
 
+#define WGPU_TARGET_WINDOWS 1
```
Comment (rajveermalviya) [inline examples/triangle/main.c:8]: `rm`

File: ffi/webgpu-headers
Comment (rajveermalviya) [inline ffi/webgpu-headers:1]: is this necessary?

File: ffi/wgpu.h
Comment (rajveermalviya) [inline ffi/wgpu.h:1]: you probably need to add an a new `WGPUBindGroupEntryExt` struct e.g:

```c
typedef struct WGPUBindGroupEntryExt {
    WGPUChainedStruct const * nextInChain;
    WGPUBuffer const * buffers; // might as well add buffer array
    size_t bufferCount;
    WGPUSampler const * samplers;
    size_t samplerCount;
    WGPUTextureView const * textureViews;
    size_t textureViewCount;
} WGPUBindGroupEntryExt;
```

and then use `follow_chain` macro in `wgpuDeviceCreateBindGroup` on entry to get the ext struct.

File: examples/triangle/main.c (hunk 1 above)
Comment (rajveermalviya) [inline examples/triangle/main.c]: rm

File: src/lib.rs [hunk 2]
```diff
@@ -1332,11 +1333,31 @@ pub unsafe extern "C" fn wgpuDeviceCreateBindGroup(
                     binding: entry.binding,
                     resource: wgc::binding_model::BindingResource::Sampler(sampler.id),
                 }
+            } else if entry.samplerArray.as_ref().is_some() {
```
Comment (rajveermalviya) [inline src/lib.rs]: you can't directly get `entry.samplerArray` as this is `webgpu.h` struct and it doesn't have this member and we can't edit `webgpu.h`.

correct way is to add a new ext struct in `wgpu.h` and then get the ext struct value using `follow_chain` macro.

File: src/lib.rs [hunk 3]
```diff
@@ -1332,11 +1333,31 @@ pub unsafe extern "C" fn wgpuDeviceCreateBindGroup(
...
+                wgc::binding_model::BindGroupEntry {
+                    binding: entry.binding,
+                    resource: wgc::binding_model::BindingResource::SamplerArray(samplers),
+                }
             } else if let Some(texture_view) = entry.textureView.as_ref() {
                 wgc::binding_model::BindGroupEntry {
                     binding: entry.binding,
                     resource: wgc::binding_model::BindingResource::TextureView(texture_view.id),
                 }
+            } else if entry.textureViewArray.as_ref().is_some() {//let Some(texture_views) = entry.textureViewArray.as_ref() {
```
Comment (rajveermalviya) [inline src/lib.rs]: same

File: src/lib.rs [hunk 4]
```diff
@@ -1489,12 +1512,14 @@ pub unsafe extern "C" fn wgpuDeviceCreateBindGroupLayout(
                 panic!("invalid bind group layout entry for bind group layout descriptor");
             };
 
+            let array_length = NonZeroU32::new(entry.count);
```
Comment (rajveermalviya) [inline src/lib.rs]: same for layout entry, you will need to create a `WGPUBindGroupLayoutEntryExt` struct in `wgpu.h`:

```c
typedef struct WGPUBindGroupLayoutEntryExt {
    WGPUChainedStruct const * nextInChain;
    uint32_t count;
} WGPUBindGroupLayoutEntryExt;
```

**Comment by Linx145** [comment]
my bad, left the define WGPU_TARGET_WINDOWS 1 in examples/triangle/main.c. Should be fixed now

**Review by rajveermalviya — DISMISSED** [review]
Thanks for contributing!

I have added some comments that needs to be addressed in order to get this merged.

File: src/lib.rs (hunk 4 above)
Comment (Linx145) [reply src/lib.rs]: Hi, sorry for inactivity, have been quite busy lately. Will get to it ASAP, thanks for the pointers!

**Comment by rajveermalviya** [comment]
Ping from triage

**Comment by rajveermalviya** [comment]
I have rebased this branch & and added a new example `texture_arrays` to demonstrate the feature, it's good to merge now!

**Review by almarklein — APPROVED** [review]

**Pull request merged**